src/ddmd/dcache.d

```diff
@@ -0,0 +1,377 @@
+            if (toc.length & (toc.length-1)) abort();
+            ourPid = getpid();
+            // Lock before modification of shared data
+            {
+                lock();
+                scope(exit) unlock();
+                // If first entry is all zeros - new mapping
+                if (data.occupied == 0 && data.size == 0)
+                {
+                    data.size = DATA_SIZE;
+                }
+            }
+            // Anything older then tsThreshold is considered ripe for eviction
+            tsCompilation = time(null);
+            tsThreshold = tsCompilation - 24*3600;
+        }
+
+        /// Clean the cache
+        void truncate()
+        {
+            lock();
+            scope(exit) unlock();
+            (cast(ubyte[])mapping)[]  = 0;
+            data.size = DATA_SIZE;
+        }
+
+        static struct TocEntry
+        {
+            uint hash32;
+            uint index;
+            ulong lastUsed; // seconds since 2000-01-01, == 0 - means entry is empty
+        }
+        static assert(TocEntry.sizeof == 16);
+
+        static struct Data
+        {
+            uint occupied; // 0 - free block, 1 - occupied
+            uint size; // size of block, that includes size of this header
+            uint next; // index from the start of data segement
+            char[0] key; // key data, zero-terminated
+            // ubyte[0] - data
+        }
+        static assert(Data.sizeof == 12);
+
+        string get(const(char)[] key)
+        {
+            lock();
+            scope(exit) unlock();
+            uint h = fnv1a(key);
+            size_t offset = h & (toc.length-1);
+            for (;;)
+            {
+                if (toc[offset].lastUsed == 0)
+                {
+                    return null;
+                }
+                if (toc[offset].hash32 == h)
+                {
+                    auto p = cast(Data*)(cast(void*)data + toc[offset].index);
+                    if (p.occupied && 0 == strncmp(p.key.ptr, key.ptr, key.length))
+                    {
+                        auto ptr = p.key.ptr + key.length + 1;
+                        return cast(string)ptr[0 .. p.size - key.length - 1 - Data.sizeof ];
+                    }
+                }
+                offset = (offset + 1) & (toc.length-1);
+            }
+        }
+
+        bool put(const(char)[] key, const(char)[] value)
+        {
+            lock();
+            scope(exit) unlock();
+            uint h = fnv1a(key);
+            size_t offset = h & (toc.length-1);
+            for (;;)
+            {
+                Data* target = cast(Data*)(cast(void*)data + toc[offset].index);
+                if (toc[offset].lastUsed == 0 || strncmp(target.key.ptr, key.ptr, key.length) == 0)
+                {
+                    //TODO: fix linear scan for free block
+                    Data* p = data;
+                    uint blockSize = cast(uint)(Data.sizeof + key.length + 1 + value.length);
```

Comment:
Oh with variable sizes entries things become a bit more complex.
Boost has allocators for shared memory, maybe sth. to draw from.
http://www.boost.org/doc/libs/1_45_0/doc/html/interprocess/allocators_containers.html#interprocess.allocators_containers.allocator_introduction.allocator_properties

You could also consider to just append new entries and sometimes run a compaction to purge overwritten/evicted values.